### ISSUE TITLE
How FasterTransformer output all hidden states and all attentions as BERT of HuggingFace?

### ISSUE THREAD

**Issue by quancq** [NVIDIA/FasterTransformer#91]
I use Pytorch code.
I found relevant code:
- [FTEncoder](https://github.com/NVIDIA/FasterTransformer/blob/main/fastertransformer/th_op/encoder.h): Line 117 assign address output to attribute of encoder_param.
- [BertEncoderTransformer](https://github.com/NVIDIA/FasterTransformer/blob/main/fastertransformer/bert_encoder_transformer.h): Line 587 is main forward function.

But I don't understand C code. Could anyone help me? Thanks.

**Comment by byshiue** [comment]
I am not sure the meaning of "all hidden states". 
If you hope to get all intermediate results during computation, I think you want to get these buffer:
* buffers of bert: from [link](https://github.com/NVIDIA/FasterTransformer/blob/main/fastertransformer/bert_encoder_transformer.h#L107) to [link](https://github.com/NVIDIA/FasterTransformer/blob/main/fastertransformer/bert_encoder_transformer.h#L110)
* buffers of attention: from [link](https://github.com/NVIDIA/FasterTransformer/blob/main/fastertransformer/cuda/open_attention.h#L205) to [link](https://github.com/NVIDIA/FasterTransformer/blob/main/fastertransformer/cuda/open_attention.h#L212).

If you only want to check value, you can dump to file by copying the data of these buffer to cpu.
If you want to set them as some ouptuts of pytorch op, then you need to modify the pytorch op.

**Comment by quancq** [comment]
Sorry, I confused with all hidden states. I used RoBERTa model from HuggingFace. In output of [RoBERTa forward](https://huggingface.co/transformers/model_doc/roberta.html#transformers.RobertaModel.forward), I need hidden_states (list num_encoder_layers of [batch_size, seq_len, hidden_size]), attentions (list num_encoder_layers of [batch_size, num_heads, seq_len, seq_len]).

My solution as below
1. Hidden_states
- Modify [Line 199 - 203](https://github.com/NVIDIA/FasterTransformer/blob/main/sample/pytorch/utils/encoder.py#L199l).
```
all_hidden_states = []
all_attr_outputs = []
for i in range(self.layer_num):            
    hidden_states, attr_output = self.encoders[i].forward(hidden_states, attention_mask, trt_seq_len, sequence_id_offset)
    all_attr_outputs.append(attr_output)
    if self.remove_padding:
        hidden_states_tmp = self.rebuild_padding(hidden_states, sequence_id_offset, attention_mask, 0)
        all_hidden_states.append(hidden_states_tmp)
return (all_hidden_states, all_attr_outputs)
```

2. Attentions
- Modify [code](https://github.com/NVIDIA/FasterTransformer/blob/main/fastertransformer/th_op/encoder.cc#L120)
```
Tensor* attr_output;
ftencoder->forward(batch_size, seq_len, input, attr_mask, output, attr_output, trt_seqlen_offset, sequence_id_offset, _remove_padding);
return output, *attr_output;
```
- Modify [forward function](https://github.com/NVIDIA/FasterTransformer/blob/main/fastertransformer/th_op/encoder.h#L94)
```
void forward(..., Tensor& output, Tensor& attr_output){
    encoder_tmp->forward();
    attr_output = encoder_tmp.attr_matmul_buf_
    encoder_tmp->freeBuffer();
}

```

@byshiue can you review my solution? Is my code right? I don't sure about C syntax and some functions/variables (Has attr_matmul_buf_ size [batch_size, seq_len, seq_len] and sum by dim=-1 is 1?; Does Tensor* attr_output use right? Does encoder_tmp->freeBuffer() free attr_matmul_buf_?) of repo.

**Comment by byshiue** [comment]
For hidden state, I think the solution is ok.
For attention, if you want to get the result of Q*K^T, then you must modify the source code of FT, but not only modify the th_op.

**Comment by byshiue** [comment]
Close this bug because it is inactivated. Feel free to re-open this issue if you still have any problem.
